src/ts/components/seekbar.ts

```diff
@@ -1074,4 +1011,130 @@ export class SeekBar extends Component<SeekBarConfig> {
+
+      if (shouldRemoveMarker(markerPosition, markerDuration)) {
+        this.removeMarker(marker);
+      } else {
+        this.timelineMarkers.push({ marker, position: markerPosition, duration: markerDuration });
+      }
+    }
+
+    // Populate the timeline with the markers
+    this.updateMarkersDOM();
+  }
+
+  private updateMarkersDOM(): void {
+    this.seekBarMarkersContainer.empty();
+
+    const seekBarWidthPx = this.seekBar.width();
+
+    for (let marker of this.timelineMarkers) {
+      const markerClasses = ['seekbar-marker'].concat(marker.marker.cssClasses || [])
+        .map(cssClass => this.prefixCss(cssClass));
+
+      const cssProperties: {[propertyName: string]: string} = {
+        'width': marker.position + '%',
+      };
+
+      if (marker.duration > 0) {
+        const markerWidthPx = Math.round(seekBarWidthPx / 100 * marker.duration);
+        cssProperties['border-right-width'] = markerWidthPx + 'px';
+        cssProperties['margin-left'] = '0';
+      }
+
+      this.seekBarMarkersContainer.append(new DOM('div', {
+        'class': markerClasses.join(' '),
+        'data-marker-time': String(marker.marker.time),
+        'data-marker-title': String(marker.marker.title),
+      }).css(cssProperties));
+    }
+  }
+  // #endregion
+}
+
+// #region Marker helper functions
+function getMarkerPositions(player: PlayerAPI, marker: TimelineMarker) {
+  const duration = getDuration(player);
+
+  const markerPosition = 100 / duration * getMarkerTime(marker, player, duration); // convert absolute time to percentage
+  let markerDuration = 100 / duration * marker.duration;
+
+  if (markerPosition < 0 && !isNaN(markerDuration)) {
+    // Shrink marker duration for on live streams as they reach end
+    markerDuration = markerDuration + markerPosition;
+  }
+
+  return { markerDuration, markerPosition };
+}
+
+function getMarkerTime (marker: TimelineMarker, player: PlayerAPI, duration: number): number {
+  if (!player.isLive()) {
+    return marker.time;
+  }
+
+  return duration - (PlayerUtils.getSeekableRange(player).end - marker.time);
+}
+
+function getDuration (player: PlayerAPI): number {
+  if (!player.isLive()) {
+    return player.getDuration();
+  }
+  const {start, end} = PlayerUtils.getSeekableRange(player);
+
+  return end - start;
+}
+
+function shouldRemoveMarker (markerPosition: number, markerDuration: number): boolean {
+  return (isNaN(markerDuration) && markerDuration < 0) || markerPosition < 0;
+}
+
+function  shouldProcessMarkers(player: PlayerAPI, uimanager: UIInstanceManager): boolean {
+  // Don't generate timeline markers if we don't yet have a duration
+  // The duration check is for buggy platforms where the duration is not available instantly (Chrome on Android 4.3)
+  const validToProcess = player.getDuration() !== Infinity || player.isLive();
+  const hasMarkers = this.uimanager.getConfig().metadata.markers.length > 0;
+
+  return validToProcess && hasMarkers;
 }
+// #endregion
```

Comment:
As a first step in refactoring this, I think it would be a good idea to move this responsibility to a delegate class.